src/main/java/graphql/nadel/engine/OverallQueryTransformer.java

```diff
@@ -282,20 +278,29 @@ QueryTransformationResult transformMergedFields(
                                                                Map<String, Object> variableValues,
                                                                Service service,
                                                                Object serviceContext) {
-        return fragments.values().stream()
-                .map(fragment -> transformFragmentDefinition(
-                        executionContext,
-                        underlyingSchema,
-                        fragment,
-                        transformationByResultField,
-                        typeRenameMappings,
-                        referencedFragmentNames,
-                        referencedVariables,
-                        serviceExecutionHooks,
-                        variableValues,
-                        service,
-                        serviceContext))
-                .collect(toMapCollector(FragmentDefinition::getName, identity()));
+
+        Set<String> fragmentsToTransform = new LinkedHashSet<>(referencedFragmentNames);
+        List<FragmentDefinition> transformedFragments = new ArrayList<>();
+        while (!fragmentsToTransform.isEmpty()) {
+            String fragmentName = fragmentsToTransform.iterator().next();
+            Set<String> newReferencedFragments = new LinkedHashSet<>();
+            FragmentDefinition transformedFragment = transformFragmentDefinition(
+                    executionContext,
+                    underlyingSchema,
+                    fragments.get(fragmentName),
+                    transformationByResultField,
+                    typeRenameMappings,
+                    newReferencedFragments,
+                    referencedVariables,
+                    serviceExecutionHooks,
+                    variableValues,
+                    service,
+                    serviceContext);
+            transformedFragments.add(transformedFragment);
+            fragmentsToTransform.addAll(newReferencedFragments);
```

Comment:
is newReferencedFragments a side effect?

Reply:
yes ... it contains the references Fragments from the transformed Fragment.